core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -181,7 +185,7 @@ public static Query create(
         result.queryManager.addOutputInfoListener(result.getQueryId(), result::setQueryOutputInfo);
 
         result.queryManager.addStateChangeListener(result.getQueryId(), state -> {
-            if (state.isDone()) {
+            if (state.isDone() || state == FINISHING) {
```

Comment:
looks like this should be separate commit.

Reply:
Or I guess I do not fully understand this change. Can you explain?

Reply:
That's a tricky one. For queries with no output stage (e.g.: DDL operations such as `DROP TABLE`) the exchange is still created. It has to be closed to allow query to finish, otherwise the client will be waiting of `ExchangeClient#isBlocked` and the query will never finish. The client is closed by the `closeExchangeClientIfNecessary`. However with this change the query will not transition to the `FINISHED` state before the results are consumed. So we have to close the `ExchangeClient` when a query is in `FINISHING` state to allow the client to pool the results and mark the query as `resultsConsumed` to allow it to finish.